base/simdloop.jl

```diff
@@ -56,7 +56,15 @@ simd_outer_range(r) = 0:0
 # Compile Expr x in context of @simd.
 function compile(x, ivdep)
     (isa(x, Expr) && x.head == :for) || throw(SimdError("for loop expected"))
-    length(x.args) == 2 || throw(SimdError("1D for loop expected"))
+    if !isa(x.args[1].args[1], Symbol)
+        # this is a multidimensional for loop. Flatten it, and apply @simd to the innermost loop
```

Comment:
I think this is overly complicated. The current `@simd` implementation only requires the insertion of `$(Expr(:loopinfo, Symbol("julia.simdloop"), ivdep))` where `ivdep` is either `nothing` or `Symbol("julia.ivdep")`. 
The loopinfo expression will only apply to the next loop-latch, e.g. the loop it is contained in.

The rest of the complexity is to efficiently handle `Cartesian Loops` which don't get lowered into a N-D loop nest, but are 1D loop over a cartesian iterator which makes it impossible for LLVM to efficiently vectorize.

Reply:
I might be confused here, but that's the point, isn't it? With this implementation (which might indeed be too complicated, I'm still finding my way around julia's macros) the last loop gets taken out and separates into a regular 1D for loop which vectorizes properly.